beaker_kernel/lib/subkernels/base.py

```diff
@@ -12,9 +26,110 @@ class BaseSubkernel(abc.ABC):
 
     WEIGHT: int = 50  # Used for auto-sorting in drop-downs, etc. Lower weights are listed earlier.
 
+
     FETCH_STATE_CODE: str = ""
 
     @classmethod
     @abc.abstractmethod
     def parse_subkernel_return(cls, execution_result) -> Any:
         ...
+
+    def __init__(self, jupyter_id: str, subkernel_configuration: dict, context):
+        self.jupyter_id = jupyter_id
+        self.connected_kernel = ProxyKernelClient(subkernel_configuration)
+        self.context = context
+    
+    def send_response(self, stream, msg_or_type, content=None, channel=None, parent_header={}, parent_identities=None):
+        return self.context.send_response(stream, msg_or_type, content, channel, parent_header, parent_identities)
+
+    async def execute(self, command, response_handler=None, parent_header={}):
+        return await self.context.execute(command, response_handler, parent_header)
+
+    async def evaluate(self, expression, parent_header={}):
+        return await self.context.evaluate(expression, parent_header)
+
+
+    def cleanup(self):
+        if self.jupyter_id is not None:
+            try:
+                print(f"Shutting down connected subkernel {self.jupyter_id}")
+                res = requests.delete(
+                    f"{server_url}/api/kernels/{self.jupyter_id}",
+                    headers={"Authorization": f"token {server_token}"},
+                )
+                if res.status_code == 204:
+                    self.jupyter_id = None
+            except requests.exceptions.HTTPError as err:
+                print(err)
+
+
+class BaseCheckpointableSubkernel(BaseSubkernel):
+    SERIALIZATION_EXTENSION: str = "storage"
+
+    def __init__(self, jupyter_id: str, subkernel_configuration: dict, context):
+        super().__init__(jupyter_id, subkernel_configuration, context)
+        self.checkpoints = None
+        self.storage_prefix = None
+
+    @property
+    def checkpoints_enabled(self):
+        return self.checkpoints is not None
+
+    def start_checkpointing(self):
+        self.checkpoints : list[Checkpoint] = []
+        self.storage_prefix = mkdtemp()
+        makedirs(self.storage_prefix, exist_ok=True)
```

Comment:
I guess I'm confused why we have this where it starts of disabled but there's a method called to enable checkpointing when a message arrives.
If the functionality is behind a feature gate, it should be either on or off, but not "maybe on or off depending on what users have done"
If it's not behind a feature gate, it should be "enabled" as part of the initialization.